### PR TITLE
Chore - possibility to inject URLSessionConfiguration to be used by ChatClient

### PR DESCRIPTION
**Pull request by mschuetz-viz**
### 🔗 Issue Links

Customer Support Ticket #39734

### 🎯 Goal

We, at Viz.ai, are heavily using `URLProtocol` based mocks for testing our application flows in our UI tests. While transitioning to Stream Chat we found that currently there's no way of injecting a prepared `URLSessionConfiguration` into the `ChatClient` to be used with our mocking engine. This pull request contains a proposal of how we could optionally inject such a prepared `URLSessionConfiguration` and unblock our development of UI tests.

### 📝 Summary
- Added a non-mutable `urlSessionConfiguration` property to the `ChatClientConfig` struct
- By default, `urlSessionConfiguration` will be initialised with `URLSessionConfiguration.default`, as before
- In `ChatClient`, `ChatClientConfig.urlSessionConfiguration` will be used instead of `URLSessionConfiguration.default`
- Added unit test to make sure that new configuration is used

### 🛠 Implementation

The goal was to change as little code as possible while making it obvious from an API perspective.

### 🎨 Showcase

not applicable - no visual changes

### 🧪 Manual Testing Notes

not applicable - default behaviour has not changed

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [x] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme
giphy-downsized

### CODE REVIEW

File: Sources/StreamChat/Config/ChatClientConfig.swift [hunk 1]
```diff
@@ -186,9 +186,17 @@ public struct ChatClientConfig {
...
-    public init(apiKey: APIKey) {
+    
+    /// The `URLSessionConfiguration` being used as default configuration for the `APIClient` and
+    /// `WebSocketClient`
+    public let urlSessionConfiguration: URLSessionConfiguration
+
+    public init(
+        apiKey: APIKey,
+        urlSessionConfiguration: URLSessionConfiguration = .default
+    ) {
```
Comment (nuno-vieira) [inline Sources/StreamChat/Config/ChatClientConfig.swift:197]: Just for consistency purposes it is better we follow the rest of the properties approach, where they are mutable with an initial value, and then you can change it whenever you create the config. 👍 

```suggestion
    /// The `URLSessionConfiguration` being used as the default configuration for the `APIClient` and
    /// `WebSocketClient`
    public var urlSessionConfiguration: URLSessionConfiguration = .default

    public init(
        apiKey: APIKey
    ) {
```

File: Sources/StreamChat/ChatClient.swift [hunk 2]
```diff
@@ -243,7 +243,7 @@ public class ChatClient {
     /// The default configuration of URLSession to be used for both the `APIClient` and `WebSocketClient`. It contains all
     /// required header auth parameters to make a successful request.
     private var urlSessionConfiguration: URLSessionConfiguration {
-        let configuration = URLSessionConfiguration.default
+        let configuration = config.urlSessionConfiguration
         configuration.waitsForConnectivity = false
         configuration.httpAdditionalHeaders = sessionHeaders
         configuration.timeoutIntervalForRequest = config.timeoutIntervalForRequest
```
Comment (nuno-vieira) [inline Sources/StreamChat/ChatClient.swift:249]: One downside here is that customers won't be able to override the `waitsForConnectivity` or the `httpAdditionalHeaders`. But for now, I think this approach is simple enough, and it should work for 99.99% of use cases. If we see the need, we can add this feature later 👍

**Review by nuno-vieira — COMMENTED** [review]
Thank you @mschuetz-viz!

We were actually going to work on it this week, but you beat us to it 😛 

The implementation was actually what we were going to do 👍 

Just added some minor comments. You don't need to change them if you don't have the time, we can do it ourselves no problem 👍 

Thank you again for your time!

File: Sources/StreamChat/Config/ChatClientConfig.swift (hunk 1 above)
Comment (mschuetz-viz) [reply Sources/StreamChat/Config/ChatClientConfig.swift]: Consistency is king 🤴 - fixed

**Comment by mschuetz-viz** [comment]
Thanks for the quick response & input @nuno-vieira - impressive work that you're doing with this SDK, quite enjoyed looking around in your code base and try figuring out a proposal for our problem.

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅ 

@mschuetz-viz Thank you again! 🙏 

We will update the changelog and merge this one once we publish the 4.36.0 release 👍

**Comment by mschuetz-viz** [comment]
Just to double check - some Github Checks are failing, is there anything I can do? Please let me know if you need anything further from my side.

**Comment by nuno-vieira** [comment]
> Just to double check - some Github Checks are failing, is there anything I can do? Please let me know if you need anything further from my side.

@mschuetz-viz It should be fine 👍 This is all permissions related things, we will handle it from here no problem 🙂

**Comment by mschuetz-viz** [comment]
@nuno-vieira is there any estimation when this will be released? no hurry, just to plan our next sprint 😅

**Comment by nuno-vieira** [comment]
Hi @mschuetz-viz!

The estimation is next week 👍

**Pull request merged**